### ISSUE TITLE
Projects and Events lose parent Collective part of URL

### ISSUE THREAD

**Issue by alanna** [opencollective/opencollective#4587]
https://www.loom.com/share/1f9a567a080e414aaf3b2bb990e6e433

If you go to a Collective with a Project and click on it, the URL will be `opencollective.com/parent/projects/projectname`. But if you look at the Project's expenses page, the URL changes to `opencollective.com/projectname/expenses`. It should be 'opencollective.com/parent/projects/projectname/expenses'.

**Comment by alanna** [comment]
Just hit this same issue again through a different path: I searched for a project name and clicked it from the search screen, and the URL I went to was opencollective.com/project instead of including the parent in the URL.

**Comment by alanna** [comment]
This keeps happening and causing confusion.

**Comment by natehn** [comment]
Just wanna note that this is coming up related to OCF's grantmaking process, in case that helps with prioritization

**Comment by alanna** [comment]
Seems this is affecting events as well... I just got linked to this one: https://opencollective.com/fair-open-data-for-public-good-and-health-2f97a527

@znarf @Betree any thoughts on how we can fix this?

**Comment by Betree** [comment]
We basically need to update all the places where we generate URLs based on accounts. I would recommend editing the helpers in https://github.com/opencollective/opencollective-frontend/blob/86d9c15b40886cadc3614ba4d5b00c7d9ff67eab/lib/url-helpers.js, refactoring, and adding new ones if necessary to reduce the amount of code needed to generate these links.

Having something like this in the code would be pretty nice:

```jsx
<>
  <Link href={getAccountUrl(account)} />
  <Link href={getAccountContributePageUrl(account)} />
  <Link href={getAccountExpensePageUrl(account)} />
</>
```

**Comment by znarf** [comment]
I would also add (and maybe do it first) a "magic" redirect mechanism for anyone visiting an URL without the parent part.

**Comment by SudharakaP** [comment]
> I would also add (and maybe do it first) a "magic" redirect mechanism for anyone visiting an URL without the parent part.

@znarf : Do you mean to say redirect for example, `/<event-slug>` to `/<parent-collective-slug>/events/<event-slug>`? I don't quite get why we need the redirect mechanism here; if the user goes to `/<event-slug>` everything works as usual. It might add unnecessary rendering as we probably need to do a GraphQl query to get the `parent-collective` when the user goes to `/<event-slug>` and then push the `/<parent-collective-slug>/events/<event-slug>` route. Am I missing something? 🤔 

cc: @Betree

**Comment by Betree** [comment]
> > I would also add (and maybe do it first) a "magic" redirect mechanism for anyone visiting an URL without the parent part.
> 
> @znarf : Do you mean to say redirect for example, `/<event-slug>` to `/<parent-collective-slug>/events/<event-slug>`? I don't quite get why we need the redirect mechanism here; if the user goes to `/<event-slug>` everything works as usual. It might add unnecessary rendering as we probably need to do a GraphQl query to get the `parent-collective` when the user goes to `/<event-slug>` and then push the `/<parent-collective-slug>/events/<event-slug>` route. Am I missing something? thinking
> 
> cc: @Betree

I'll let @znarf comment, but I think it's about having consistent routes and the fact that this will catch 100% of the situations - even if we forget to update a link somewhere. It can be done without impacting the user experience by using [shallow routing](https://nextjs.org/docs/routing/shallow-routing).

Also important for this issue: we should make sure that a `canonicalUrl` is properly set in all meta, that's how we're going to make sure search engines update their content.

**Comment by znarf** [comment]
> Do you mean to say redirect for example, /<event-slug> to /<parent-collective-slug>/events/<event-slug>? 

Yes.

> I don't quite get why we need the redirect mechanism here; if the user goes to /<event-slug> everything works as usual. 

We want to redirect to the canonical URL, because this is the one they will be sharing (copy/pasting) with other people. It's one elegant way to progressively get rid of the non-canonical URLs while we chase edge cases.

**Comment by SudharakaP** [comment]
Thanks for the clarifications. 

@Betree : I've set the meta here; 

https://github.com/opencollective/opencollective-frontend/blob/f0bd38f23f0af816ad7fa5bb39d56c8da7093d96/pages/collective-page.js#L174-L176

Let me know if there's anything else to be added to the meta. 🤔 

I've redirected all the places to the canonical URL as suggested using the shallow routing. Let me know if there's anything missing or needs to be changed (screen-cast; https://github.com/opencollective/opencollective-frontend/pull/7304#issue-1094033956).
